### ISSUE TITLE
Invalid plan produced for this query

### ISSUE THREAD

**Issue by haozhun** [facebook/presto#3714]
`EXPLAIN SELECT APPROX_PERCENTILE(0.5,x) med_x, AVG(x) avg_x FROM (values 1, 2, 3) t(x);`

```
java.lang.IllegalArgumentException: Invalid node. Aggregation dependencies ([field_1]) not in source plan output ([expr, x])
	at com.google.common.base.Preconditions.checkArgument(Preconditions.java:145)
	at com.facebook.presto.sql.planner.PlanSanityChecker.checkDependencies(PlanSanityChecker.java:485)
	at com.facebook.presto.sql.planner.PlanSanityChecker.access$100(PlanSanityChecker.java:62)
	at com.facebook.presto.sql.planner.PlanSanityChecker$Visitor.visitAggregation(PlanSanityChecker.java:99)
	at com.facebook.presto.sql.planner.PlanSanityChecker$Visitor.visitAggregation(PlanSanityChecker.java:71)
	at com.facebook.presto.sql.planner.plan.AggregationNode.accept(AggregationNode.java:160)
	at com.facebook.presto.sql.planner.PlanSanityChecker$Visitor.visitProject(PlanSanityChecker.java:208)
	at com.facebook.presto.sql.planner.PlanSanityChecker$Visitor.visitProject(PlanSanityChecker.java:71)
	at com.facebook.presto.sql.planner.plan.ProjectNode.accept(ProjectNode.java:81)
	at com.facebook.presto.sql.planner.PlanSanityChecker$Visitor.visitProject(PlanSanityChecker.java:208)
	at com.facebook.presto.sql.planner.PlanSanityChecker$Visitor.visitProject(PlanSanityChecker.java:71)
	at com.facebook.presto.sql.planner.plan.ProjectNode.accept(ProjectNode.java:81)
	at com.facebook.presto.sql.planner.PlanSanityChecker$Visitor.visitProject(PlanSanityChecker.java:208)
	at com.facebook.presto.sql.planner.PlanSanityChecker$Visitor.visitProject(PlanSanityChecker.java:71)
	at com.facebook.presto.sql.planner.plan.ProjectNode.accept(ProjectNode.java:81)
	at com.facebook.presto.sql.planner.PlanSanityChecker$Visitor.visitProject(PlanSanityChecker.java:208)
	at com.facebook.presto.sql.planner.PlanSanityChecker$Visitor.visitProject(PlanSanityChecker.java:71)
	at com.facebook.presto.sql.planner.plan.ProjectNode.accept(ProjectNode.java:81)
	at com.facebook.presto.sql.planner.PlanSanityChecker$Visitor.visitOutput(PlanSanityChecker.java:258)
	at com.facebook.presto.sql.planner.PlanSanityChecker$Visitor.visitOutput(PlanSanityChecker.java:71)
	at com.facebook.presto.sql.planner.plan.OutputNode.accept(OutputNode.java:81)
	at com.facebook.presto.sql.planner.PlanSanityChecker.validate(PlanSanityChecker.java:68)
	at com.facebook.presto.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:94)
	at com.facebook.presto.sql.analyzer.QueryExplainer.getLogicalPlan(QueryExplainer.java:134)
	at com.facebook.presto.sql.analyzer.QueryExplainer.getPlan(QueryExplainer.java:91)
	at com.facebook.presto.sql.analyzer.StatementAnalyzer.getQueryPlan(StatementAnalyzer.java:587)
	at com.facebook.presto.sql.analyzer.StatementAnalyzer.visitExplain(StatementAnalyzer.java:568)
	at com.facebook.presto.sql.analyzer.StatementAnalyzer.visitExplain(StatementAnalyzer.java:124)
	at com.facebook.presto.sql.tree.Explain.accept(Explain.java:54)
	at com.facebook.presto.sql.tree.AstVisitor.process(AstVisitor.java:22)
	at com.facebook.presto.sql.analyzer.Analyzer.analyze(Analyzer.java:60)
	at com.facebook.presto.execution.SqlQueryExecution.doAnalyzeQuery(SqlQueryExecution.java:259)
	at com.facebook.presto.execution.SqlQueryExecution.analyzeQuery(SqlQueryExecution.java:245)
	at com.facebook.presto.execution.SqlQueryExecution.start(SqlQueryExecution.java:209)
	at com.facebook.presto.execution.QueuedExecution.lambda$start$117(QueuedExecution.java:68)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

**Comment by raghavsethi** [comment]
The second argument of `approx_percentile` should be the percentile, not the first. So this query *should* fail, but with this: `com.facebook.presto.spi.PrestoException: Percentile must be between 0 and 1`.

**Comment by raghavsethi** [comment]
It's interesting that the ordering of expressions also matters:
```
    SELECT AVG(x), APPROX_PERCENTILE(0.5,x) FROM (values 1, 2, 3) t(x);
```
results in a casting error:

```
Query 20151015_184910_00019_ys3ku failed: com.facebook.presto.sql.tree.Cast cannot be cast to com.facebook.presto.sql.tree.QualifiedNameReference
java.lang.ClassCastException: com.facebook.presto.sql.tree.Cast cannot be cast to com.facebook.presto.sql.tree.QualifiedNameReference
	at com.facebook.presto.sql.planner.LocalExecutionPlanner$Visitor.buildAccumulatorFactory(LocalExecutionPlanner.java:1656)
	at com.facebook.presto.sql.planner.LocalExecutionPlanner$Visitor.planGlobalAggregation(LocalExecutionPlanner.java:1682)
	at com.facebook.presto.sql.planner.LocalExecutionPlanner$Visitor.visitAggregation(LocalExecutionPlanner.java:687)
	at com.facebook.presto.sql.planner.LocalExecutionPlanner$Visitor.visitAggregation(LocalExecutionPlanner.java:410)
	at com.facebook.presto.sql.planner.plan.AggregationNode.accept(AggregationNode.java:160)
	at com.facebook.presto.sql.planner.LocalExecutionPlanner$Visitor.visitOutput(LocalExecutionPlanner.java:433)
	at com.facebook.presto.sql.planner.LocalExecutionPlanner$Visitor.visitOutput(LocalExecutionPlanner.java:410)
	at com.facebook.presto.sql.planner.plan.OutputNode.accept(OutputNode.java:81)
	at com.facebook.presto.sql.planner.LocalExecutionPlanner.plan(LocalExecutionPlanner.java:242)
	at com.facebook.presto.execution.SqlTaskExecution.<init>(SqlTaskExecution.java:165)
	at com.facebook.presto.execution.SqlTaskExecution.createSqlTaskExecution(SqlTaskExecution.java:114)
	at com.facebook.presto.execution.SqlTaskExecutionFactory.create(SqlTaskExecutionFactory.java:73)
	at com.facebook.presto.execution.SqlTask.updateTask(SqlTask.java:250)
...
```

**Comment by raghavsethi** [comment]
Merged #3772.
